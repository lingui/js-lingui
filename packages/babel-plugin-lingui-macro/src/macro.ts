import { createMacro, MacroParams } from "babel-plugin-macros"

import type { VisitNodeObject } from "@babel/traverse"
import { Program } from "@babel/types"

import linguiPlugin from "./index"
import { JsMacroName, JsxMacroName } from "./constants"

function macro({ state, babel, config }: MacroParams) {
  if (!state.get("linguiProcessed")) {
    state.opts = config
    const plugin = linguiPlugin(babel)

    const { enter, exit } = plugin.visitor.Program as VisitNodeObject<
      any,
      Program
    >

    enter(state.file.path, state)
    state.file.path.traverse(plugin.visitor, state)
    exit(state.file.path, state)

    state.set("linguiProcessed", true)
  }

  return { keepImports: true }
}

;[
  JsMacroName.defineMessage,
  JsMacroName.msg,
  JsMacroName.t,
  JsMacroName.useLingui,
  JsMacroName.plural,
  JsMacroName.select,
  JsMacroName.selectOrdinal,

  JsxMacroName.Trans,
  JsxMacroName.Plural,
  JsxMacroName.Select,
  JsxMacroName.SelectOrdinal,
].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throw new Error(
        `The macro you imported from "@lingui/core/macro" or "@lingui/react/macro" is being executed outside the context of compilation with babel-plugin-macros. ` +
          `This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. ` +
          `Please see the documentation for how to configure babel-plugin-macros properly: ` +
          "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md"
      )
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
}) as { isBabelMacro: true }
