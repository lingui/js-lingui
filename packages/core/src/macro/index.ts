import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJs } from "./macroJs"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export type LinguiMacroOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

export const jsMacroTags = new Set([
  "defineMessage",
  "arg",
  "t",
  "plural",
  "select",
  "selectOrdinal",
])

let config: LinguiConfigNormalized

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { i18nImportModule, i18nImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsNodes = new Set<NodePath>()
  let needsI18nImport = false

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsMacroTags.has(tagName)) {
      nodes.forEach((node) => {
        jsNodes.add(node.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsNodesArray = Array.from(jsNodes)

  jsNodesArray.filter(isRootPath(jsNodesArray)).forEach((path) => {
    const macro = new MacroJs(babel, { i18nImportName, stripNonEssentialProps })
    if (macro.replacePath(path)) needsI18nImport = true
  })

  if (needsI18nImport) {
    addImport(
      babel.types,
      state.file.path.node,
      i18nImportModule,
      i18nImportName
    )
  }
}

;[...jsMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/core/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
