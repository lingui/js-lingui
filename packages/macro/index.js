module.exports = require("@lingui/babel-plugin-lingui-macro/macro")
