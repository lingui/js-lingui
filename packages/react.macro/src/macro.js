import { createMacro } from "babel-plugin-macros"
import { implementationPlugin as TransformReactPlugin } from "@lingui/babel-plugin-transform-react"

const importsToCarryOver = ["DateFormat", "NumberFormat"]

function macro({ references, state, babel }) {
  const { types: t } = babel
  const plugin = new TransformReactPlugin(babel)

  const toKeepImports = ["Trans"]

  // Trick the plugin into thinking we've processed an import
  plugin.setImportDeclarations(
    Object.keys(references).reduce((obj, key) => {
      if (key !== "default" || key !== "Trans") obj[key] = key
      return obj
    }, {})
  )

  for (let [tagName, tags] of Object.entries(references)) {
    if (importsToCarryOver.includes(tagName)) toKeepImports.push(tagName)

    tags.forEach(openingTag => {
      if (!t.isJSXOpeningElement(openingTag.container)) return // Exclude closing elements

      const node = openingTag.context.parentPath.container

      plugin.JSXElement({ node }, state.file)
    })
  }

  const linguiReactImport = state.file.path.node.body.find(
    importNode =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === "@lingui/react"
  )

  // Handle adding the import or altering the existing import
  if (linguiReactImport) {
    toKeepImports.forEach(name => {
      if (
        linguiReactImport.specifiers.findIndex(
          specifier => specifier.imported && specifier.imported.name === name
        ) === -1
      ) {
        linguiReactImport.specifiers.push(
          t.importSpecifier(t.identifier(name), t.identifier(name))
        )
      }
    })
  } else {
    state.file.path.node.body.unshift(
      t.importDeclaration(
        toKeepImports.map(name =>
          t.importSpecifier(t.identifier(name), t.identifier(name))
        ),
        t.stringLiteral("@lingui/react")
      )
    )
  }
}

export default createMacro(macro)
// For intellisense as these will be ignored by babel-macro
// Might want to copy PropTypes here?
export const Trans = () => {}
export const Plural = () => {}
export const Select = () => {}
export const SelectOrdinal = () => {}
export const DateFormat = () => {}
export const NumberFormat = () => {}
