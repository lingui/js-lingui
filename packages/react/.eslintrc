{
  "extends": ["../../.eslintrc", "plugin:react/recommended", "plugin:react-hooks/recommended"],
  "settings": {
    "react": {
      "version": "detect" // "detect" automatically picks the version of React that is installed
    }
  },
  "rules": {
    "react/prop-types": "off",
    "react/display-name": "off"
  }
}
