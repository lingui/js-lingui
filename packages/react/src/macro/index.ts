import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJSX } from "./macroJsx"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  LinguiMacroOpts,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export const jsxMacroTags = new Set([
  "Trans",
  "Plural",
  "Select",
  "SelectOrdinal",
])

let config: LinguiConfigNormalized

// todo, make memoization on the conf library level
function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { TransImportModule, TransImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsxNodes = new Set<NodePath>()

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsxMacroTags.has(tagName)) {
      // babel-plugin-macros return JSXIdentifier nodes.
      // Which is for every JSX element would be presented twice (opening / close)
      // Here we're taking JSXElement and dedupe it.
      nodes.forEach((node) => {
        // identifier.openingElement.jsxElement
        jsxNodes.add(node.parentPath.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsxNodesArray = Array.from(jsxNodes)

  jsxNodesArray.filter(isRootPath(jsxNodesArray)).forEach((path) => {
    const macro = new MacroJSX(babel, { stripNonEssentialProps })
    macro.replacePath(path)
  })

  if (jsxNodes.size) {
    addImport(
      babel.types,
      state.file.path.node,
      TransImportModule,
      TransImportName
    )
  }
}

;[...jsxMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/react/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
