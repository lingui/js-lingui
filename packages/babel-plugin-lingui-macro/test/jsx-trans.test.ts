import { macroTester } from "./macroTester"
describe.skip("", () => {})

macroTester({
  cases: [
    {
      name: "Generate ID from message",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>Hello World</Trans>;
      `,
    },
    {
      name: "Generate different id when context provided",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>Hello World</Trans>;
        <Trans context="my context">Hello World</Trans>;
      `,
    },
    {
      name: "Preserve custom ID (string literal)",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans id="msg.hello">Hello World</Trans>;
      `,
    },
    {
      name: "Preserve custom ID (literal expression)",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans id={"msg.hello"}>Hello World</Trans>;
      `,
    },
    {
      name: "Preserve custom ID (template expression)",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans id={\`msg.hello\`}>Hello World</Trans>;
      `,
    },
    {
      name: "Should preserve reserved props: `comment`, `context`, `render`, `id`",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans
          comment="Comment for translator"
          context="translation context"
          id="custom.id"
          render={() => {}}
        >Hello World</Trans>;
      `,
    },
    {
      name: "Trans macro could be renamed",
      code: `
        import { Trans as Trans2 } from '@lingui/react/macro';
        <Trans2>Hello World</Trans2>;
      `,
    },
    {
      name: "Variables are converted to named arguments",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>Hi {yourName}, my name is {myName}</Trans>;
      `,
    },
    {
      name: "Variables are deduplicated",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>{duplicate} variable {duplicate}</Trans>;
      `,
    },
    {
      name: "Quoted JSX attributes are handled",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>Speak "friend"!</Trans>;
        <Trans id="custom-id">Speak "friend"!</Trans>;
      `,
    },
    {
      name: "HTML attributes are handled",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>
          <Text>This should work &nbsp;</Text>
        </Trans>;
      `,
    },
    {
      name: "Template literals as children",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>{\`How much is \${expression}? \${count}\`}</Trans>;
      `,
    },
    {
      name: "Strings as children are preserved",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>{"hello {count, plural, one {world} other {worlds}}"}</Trans>;
      `,
    },
    {
      name: "Expressions are converted to positional arguments",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>
          Property {props.name},
          function {random()},
          array {array[index]},
          constant {42},
          object {new Date()},
          everything {props.messages[index].value()}
        </Trans>;
      `,
    },
    {
      name: "JSX Macro inside JSX conditional expressions",
      code: `
       import { Trans } from '@lingui/react/macro';
       <Trans>Hello, {props.world ? <Trans>world</Trans> : <Trans>guys</Trans>}</Trans>
      `,
    },
    {
      name: "JSX Macro inside JSX multiple nested conditional expressions",
      code: `
      import { Trans } from '@lingui/react/macro';
      <Trans>Hello, {props.world ? <Trans>world</Trans> : (
        props.b
          ? <Trans>nested</Trans>
          : <Trans>guys</Trans>
      )
    }</Trans>
      `,
    },
    {
      name: "Elements are replaced with placeholders",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>
          Hello <strong>World!</strong><br />
          <p>
            My name is <a href="/about">{" "}
            <em>{name}</em></a>
          </p>
        </Trans>;
      `,
    },
    {
      name: "Elements inside expression container",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>{<span>Component inside expression container</span>}</Trans>;
      `,
    },
    {
      name: "Elements without children",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>{<br />}</Trans>;
      `,
    },
    {
      name: "Production - only essential props are kept",
      production: true,
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans id="msg.hello" context="my context" comment="Hello World">Hello World</Trans>
      `,
    },
    {
      name: "Production - all props kept if extract: true",
      production: true,
      macroOpts: {
        extract: true,
      },
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans id="msg.hello" comment="Hello World">Hello World</Trans>
      `,
    },
    {
      name: "Production - import type doesn't interference on normal import",
      production: true,
      useTypescriptPreset: true,
      code: `
        import type { withI18nProps } from '@lingui/react'
        import { Trans } from '@lingui/react/macro';
        <Trans id="msg.hello" comment="Hello World">Hello World</Trans>
      `,
    },
    {
      name: "Strip whitespace around arguments",
      code: `
        import { Trans } from "@lingui/react/macro";
        <Trans>
          Strip whitespace around arguments: '
          {name}
          '
        </Trans>
      `,
    },
    {
      name: "Strip whitespace around tags but keep forced spaces",
      code: `
        import { Trans } from "@lingui/react/macro";
        <Trans>
          Strip whitespace around tags, but keep{" "}
          <strong>forced spaces</strong>
          !
        </Trans>
      `,
    },
    {
      name: "Strip whitespace around tags but keep whitespaces in JSX containers",
      code: `
      import { Trans } from "@lingui/react/macro";
        <Trans>
        {"Wonderful framework "}
        <a href="https://nextjs.org">Next.js</a>
        {" say hi. And "}
        <a href="https://nextjs.org">Next.js</a>
        {" say hi."}
      </Trans>
      `,
    },
    {
      name: "Keep forced newlines",
      filename: "./jsx-keep-forced-newlines.js",
    },
    {
      name: "Use a js macro inside a JSX Attribute of a component handled by JSX macro",
      code: `
        import { Trans } from '@lingui/react/macro';
        import { t } from '@lingui/core/macro';
        <Trans>Read <a href="/more" title={t\`Full content of \${articleName}\`}>more</a></Trans>
      `,
    },
    {
      name: "Use a js macro inside a JSX Attribute of a non macro JSX component",
      code: `
        import { plural } from '@lingui/core/macro';
        <a href="/about" title={plural(count, {
          one: "# book",
          other: "# books"
        })}>About</a>
      `,
    },
    {
      name: "Ignore JSXEmptyExpression",
      code: `
        import { Trans } from '@lingui/react/macro';
        <Trans>Hello {/* and I cannot stress this enough */} World</Trans>;
      `,
    },
    {
      name: "Use decoded html entities",
      code: `
        import { Trans } from "@lingui/react/macro";
        <Trans>&amp;</Trans>
      `,
    },
    {
      name: "Should not process non JSXElement nodes",
      useTypescriptPreset: true,
      code: `
        import { Trans } from "@lingui/react/macro";
        type X = typeof Trans;
        const cmp = <Trans>Hello</Trans>
      `,
    },
  ],
})
