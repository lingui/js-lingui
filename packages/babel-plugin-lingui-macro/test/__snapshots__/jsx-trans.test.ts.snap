// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements are replaced with placeholders 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Hello <strong>World!</strong>
  <br />
  <p>
    My name is{" "}
    <a href="/about">
      {" "}
      <em>{name}</em>
    </a>
  </p>
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "k9gsHO",
      message: "Hello <0>World!</0><1/><2>My name is <3> <4>{name}</4></3></2>",
      values: {
        name: name,
      },
      components: {
        0: <strong />,
        1: <br />,
        2: <p />,
        3: <a href="/about" />,
        4: <em />,
      },
    }
  }
/>;

`;

exports[`Elements inside expression container 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>{<span>Component inside expression container</span>}</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "1cZQQW",
      message: "<0>Component inside expression container</0>",
      components: {
        0: <span />,
      },
    }
  }
/>;

`;

exports[`Elements without children 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>{<br />}</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "SCJtqt",
      message: "<0/>",
      components: {
        0: <br />,
      },
    }
  }
/>;

`;

exports[`Expressions are converted to positional arguments 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Property {props.name}, function {random()}, array {array[index]}, constant{" "}
  {42}, object {new Date()}, everything {props.messages[index].value()}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "HjKDmx",
      message:
        "Property {0}, function {1}, array {2}, constant {3}, object {4}, everything {5}",
      values: {
        0: props.name,
        1: random(),
        2: array[index],
        3: 42,
        4: new Date(),
        5: props.messages[index].value(),
      },
    }
  }
/>;

`;

exports[`Generate ID from message 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>Hello World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "mY42CM",
      message: "Hello World",
    }
  }
/>;

`;

exports[`Generate different id when context provided 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>Hello World</Trans>;
<Trans context="my context">Hello World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "mY42CM",
      message: "Hello World",
    }
  }
/>;
<_Trans
  {
    /*i18n*/
    ...{
      id: "SO/WB8",
      message: "Hello World",
      context: "my context",
    }
  }
/>;

`;

exports[`HTML attributes are handled 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  <Text>This should work &nbsp;</Text>
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "K/1Xpr",
      message: "<0>This should work \\xA0</0>",
      components: {
        0: <Text />,
      },
    }
  }
/>;

`;

exports[`Ignore JSXEmptyExpression 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>Hello {/* and I cannot stress this enough */} World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "i0M2R8",
      message: "Hello  World",
    }
  }
/>;

`;

exports[`JSX Macro inside JSX conditional expressions 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Hello, {props.world ? <Trans>world</Trans> : <Trans>guys</Trans>}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "UT5PlM",
      message: "Hello, {0}",
      values: {
        0: props.world ? (
          <_Trans
            {
              /*i18n*/
              ...{
                id: "ELi2P3",
                message: "world",
              }
            }
          />
        ) : (
          <_Trans
            {
              /*i18n*/
              ...{
                id: "39nd+2",
                message: "guys",
              }
            }
          />
        ),
      },
    }
  }
/>;

`;

exports[`JSX Macro inside JSX multiple nested conditional expressions 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Hello,{" "}
  {props.world ? (
    <Trans>world</Trans>
  ) : props.b ? (
    <Trans>nested</Trans>
  ) : (
    <Trans>guys</Trans>
  )}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "UT5PlM",
      message: "Hello, {0}",
      values: {
        0: props.world ? (
          <_Trans
            {
              /*i18n*/
              ...{
                id: "ELi2P3",
                message: "world",
              }
            }
          />
        ) : props.b ? (
          <_Trans
            {
              /*i18n*/
              ...{
                id: "lV+268",
                message: "nested",
              }
            }
          />
        ) : (
          <_Trans
            {
              /*i18n*/
              ...{
                id: "39nd+2",
                message: "guys",
              }
            }
          />
        ),
      },
    }
  }
/>;

`;

exports[`Preserve custom ID (literal expression) 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans id={"msg.hello"}>Hello World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
      message: "Hello World",
    }
  }
/>;

`;

exports[`Preserve custom ID (string literal) 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans id="msg.hello">Hello World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
      message: "Hello World",
    }
  }
/>;

`;

exports[`Preserve custom ID (template expression) 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans id={\`msg.hello\`}>Hello World</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
      message: "Hello World",
    }
  }
/>;

`;

exports[`Production - all props kept if extract: true 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans id="msg.hello" comment="Hello World">
  Hello World
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
      message: "Hello World",
      comment: "Hello World",
    }
  }
/>;

`;

exports[`Production - import type doesn't interference on normal import 1`] = `
import type { withI18nProps } from "@lingui/react";
import { Trans } from "@lingui/react/macro";
<Trans id="msg.hello" comment="Hello World">
  Hello World
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
    }
  }
/>;

`;

exports[`Production - only essential props are kept 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans id="msg.hello" context="my context" comment="Hello World">
  Hello World
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "msg.hello",
    }
  }
/>;

`;

exports[`Quoted JSX attributes are handled 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>Speak "friend"!</Trans>;
<Trans id="custom-id">Speak "friend"!</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "NWmRwM",
      message: 'Speak "friend"!',
    }
  }
/>;
<_Trans
  {
    /*i18n*/
    ...{
      id: "custom-id",
      message: 'Speak "friend"!',
    }
  }
/>;

`;

exports[`Should not process non JSXElement nodes 1`] = `
import { Trans } from "@lingui/react/macro";
type X = typeof Trans;
const cmp = <Trans>Hello</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
const cmp = (
  <_Trans
    {
      /*i18n*/
      ...{
        id: "uzTaYi",
        message: "Hello",
      }
    }
  />
);

`;

exports[`Should preserve reserved props: \`comment\`, \`context\`, \`render\`, \`id\` 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans
  comment="Comment for translator"
  context="translation context"
  id="custom.id"
  render={() => {}}
>
  Hello World
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  render={() => {}}
  {
    /*i18n*/
    ...{
      id: "custom.id",
      message: "Hello World",
      comment: "Comment for translator",
      context: "translation context",
    }
  }
/>;

`;

exports[`Strings as children are preserved 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>{"hello {count, plural, one {world} other {worlds}}"}</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "U8dd/d",
      message: "hello {count, plural, one {world} other {worlds}}",
    }
  }
/>;

`;

exports[`Strip whitespace around arguments 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>Strip whitespace around arguments: '{name}'</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "tRMgLt",
      message: "Strip whitespace around arguments: '{name}'",
      values: {
        name: name,
      },
    }
  }
/>;

`;

exports[`Strip whitespace around tags but keep forced spaces 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Strip whitespace around tags, but keep <strong>forced spaces</strong>!
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "Ud4KOf",
      message: "Strip whitespace around tags, but keep <0>forced spaces</0>!",
      components: {
        0: <strong />,
      },
    }
  }
/>;

`;

exports[`Strip whitespace around tags but keep whitespaces in JSX containers 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  {"Wonderful framework "}
  <a href="https://nextjs.org">Next.js</a>
  {" say hi. And "}
  <a href="https://nextjs.org">Next.js</a>
  {" say hi."}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "3YVd0H",
      message:
        "Wonderful framework <0>Next.js</0> say hi. And <1>Next.js</1> say hi.",
      components: {
        0: <a href="https://nextjs.org" />,
        1: <a href="https://nextjs.org" />,
      },
    }
  }
/>;

`;

exports[`Template literals as children 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>{\`How much is \${expression}? \${count}\`}</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "exe3kM",
      message: "How much is {expression}? {count}",
      values: {
        expression: expression,
        count: count,
      },
    }
  }
/>;

`;

exports[`Trans macro could be renamed 1`] = `
import { Trans as Trans2 } from "@lingui/react/macro";
<Trans2>Hello World</Trans2>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "mY42CM",
      message: "Hello World",
    }
  }
/>;

`;

exports[`Use a js macro inside a JSX Attribute of a component handled by JSX macro 1`] = `
import { Trans } from "@lingui/react/macro";
import { t } from "@lingui/core/macro";
<Trans>
  Read{" "}
  <a href="/more" title={t\`Full content of \${articleName}\`}>
    more
  </a>
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { i18n as _i18n } from "@lingui/core";
import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "QZyANg",
      message: "Read <0>more</0>",
      components: {
        0: (
          <a
            href="/more"
            title={_i18n._(
              /*i18n*/
              {
                id: "qzc3IN",
                message: "Full content of {articleName}",
                values: {
                  articleName: articleName,
                },
              }
            )}
          />
        ),
      },
    }
  }
/>;

`;

exports[`Use a js macro inside a JSX Attribute of a non macro JSX component 1`] = `
import { plural } from "@lingui/core/macro";
<a
  href="/about"
  title={plural(count, {
    one: "# book",
    other: "# books",
  })}
>
  About
</a>;

↓ ↓ ↓ ↓ ↓ ↓

import { i18n as _i18n } from "@lingui/core";
<a
  href="/about"
  title={_i18n._(
    /*i18n*/
    {
      id: "esnaQO",
      message: "{count, plural, one {# book} other {# books}}",
      values: {
        count: count,
      },
    }
  )}
>
  About
</a>;

`;

exports[`Use decoded html entities 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>&amp;</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "EwTON7",
      message: "&",
    }
  }
/>;

`;

exports[`Variables are converted to named arguments 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  Hi {yourName}, my name is {myName}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "y10VRI",
      message: "Hi {yourName}, my name is {myName}",
      values: {
        yourName: yourName,
        myName: myName,
      },
    }
  }
/>;

`;

exports[`Variables are deduplicated 1`] = `
import { Trans } from "@lingui/react/macro";
<Trans>
  {duplicate} variable {duplicate}
</Trans>;

↓ ↓ ↓ ↓ ↓ ↓

import { Trans as _Trans } from "@lingui/react";
<_Trans
  {
    /*i18n*/
    ...{
      id: "+nhkwg",
      message: "{duplicate} variable {duplicate}",
      values: {
        duplicate: duplicate,
      },
    }
  }
/>;

`;
